cpu/atxmega/periph/timer.c

```diff
@@ -0,0 +1,226 @@
+    if (tim >= TIMER_NUMOF) {
+        return -1;
+    }
+
+    /* figure out if freq is applicable */
+    for (; pre < PRESCALE_NUMOF; pre++) {
+        if ((CLOCK_CORECLOCK >> prescalers[pre]) == freq) {
+            break;
+        }
+    }
+    if (pre == PRESCALE_NUMOF) {
+        DEBUG("timer.c: prescaling failed!\n");
+        return -1;
+    }
+
+    /* stop and reset timer */
+    ctx[tim].dev->CTRLA = 0;                    /* Stop */
+    ctx[tim].dev->CTRLFSET = TC_CMD_RESET_gc;   /* Force Reset */
+
+    /* save interrupt context and timer Prescaler */
+    ctx[tim].cb = cb;
+    ctx[tim].arg = arg;
+    ctx[tim].mode = (0x0F & (pre + 1));
+
+    /* enable timer with calculated prescaler */
+    ctx[tim].dev->CTRLA = ctx[tim].mode;
+    DEBUG("timer.c: prescaler set to %d \n", ctx[tim].dev->CTRLA );
+
+    return 0;
+}
+
+int timer_set_absolute(tim_t tim, int channel, unsigned int value)
+{
+    if (channel >= CHANNELS) {
+        return -1;
+    }
+    DEBUG("timer_set_absolute channel %d to %u\n", channel, value );
+
+    // Compare or Capture Disable
+    ctx[tim].dev->CTRLB &= ~(1 << (channel + TC0_CCAEN_bp));
+
+    // Clear Interrupt Flag
+    ctx[tim].dev->INTFLAGS &= ~(1 << (channel + TC0_CCAIF_bp));
+
+    // set value to compare
+    *(((uint16_t *)(&ctx[tim].dev->CCA)) + channel) = (uint16_t)value;
+
+    /* Compare or Capture Interrupt Level */
+    ctx[tim].dev->INTCTRLB |= (ctx[tim].int_lvl << channel);
+
+    // Compare or Capture Enable
+    ctx[tim].dev->CTRLB |= (1 << (channel + TC0_CCAEN_bp));
+
+    return 1;
+}
+
+int timer_clear(tim_t tim, int channel)
+{
+    if (channel >= CHANNELS) {
+        return -1;
+    }
+
+    DEBUG("timer_clear channel %d\n", channel  );
+
+    // Compare or Capture Disable
+    ctx[tim].dev->CTRLB &= ~(1 << (channel + TC0_CCAEN_bp));
+
+    /* Clear Interrupt Flag
+     * The CCxIF is automatically cleared when the corresponding
+     * interrupt vector is executed.*/
+
+    return 0;
+}
+
+unsigned int timer_read(tim_t tim)
+{
+    DEBUG("timer_read\n");
+    return (unsigned int)ctx[tim].dev->CNT;
+}
+
+void timer_stop(tim_t tim)
+{
+    DEBUG("timer_stop\n");
+    ctx[tim].dev->CTRLA = 0;
+}
+
+void timer_start(tim_t tim)
+{
+    DEBUG("timer_start\n");
+    ctx[tim].dev->CTRLA = ctx[tim].mode;
+}
+
+#ifdef TIMER_NUMOF
+static inline void _isr(tim_t tim, int channel)
+{
+    avr8_enter_isr();
+
+    DEBUG("timer _isr channel %d\n", channel);
+
+    /* Compare or Capture Disable */
+    ctx[tim].dev->CTRLB &= ~(1 << (channel + TC0_CCAEN_bp));
+
+    /* Execute callback */
+    ctx[tim].cb(ctx[tim].arg, channel);
+
+    /* Initiate context switch if necessary */
+    if (sched_context_switch_request) {
+        thread_yield();
+    }
+
+    avr8_exit_isr();
+}
+#endif
+
+#ifdef TIMER_0
+ISR(TIMER_0_ISRA, ISR_BLOCK)
+{
+    _isr(0, 0);
+}
+ISR(TIMER_0_ISRB, ISR_BLOCK)
+{
+    _isr(0, 1);
+}
+ISR(TIMER_0_OVF, ISR_BLOCK)
+{
+    // _isr(0, 2);
```

Comment:
missed that one?

Reply:
Initially I not invest too much time in timers.
It needs a re-work.